engine/access/rpc/backend/script_executor_test.go

```diff
@@ -0,0 +1,288 @@
+	"github.com/onflow/flow-go/engine/execution/testutil"
+	"github.com/onflow/flow-go/fvm"
+	"github.com/onflow/flow-go/fvm/storage/derived"
+	"github.com/onflow/flow-go/fvm/storage/snapshot"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/execution"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/metrics"
+	"github.com/onflow/flow-go/module/state_synchronization/indexer"
+	syncmock "github.com/onflow/flow-go/module/state_synchronization/mock"
+	synctest "github.com/onflow/flow-go/module/state_synchronization/requester/unittest"
+	"github.com/onflow/flow-go/storage"
+	storageMock "github.com/onflow/flow-go/storage/mock"
+	pebbleStorage "github.com/onflow/flow-go/storage/pebble"
+	"github.com/onflow/flow-go/utils/unittest"
+	"github.com/onflow/flow-go/utils/unittest/mocks"
+)
+
+type ScriptExecutorSuite struct {
+	suite.Suite
+
+	log            zerolog.Logger
+	registerIndex  storage.RegisterIndex
+	versionControl *version.VersionControl
+	reporter       *syncmock.IndexReporter
+	indexReporter  *index.Reporter
+	scripts        *execution.Scripts
+	chain          flow.Chain
+	dbDir          string
+	height         uint64
+	headers        storage.Headers
+	vm             *fvm.VirtualMachine
+	vmCtx          fvm.Context
+	snapshot       snapshot.SnapshotTree
+}
+
+func TestScriptExecutorSuite(t *testing.T) {
+	suite.Run(t, new(ScriptExecutorSuite))
+}
+
+func newBlockHeadersStorage(blocks []*flow.Block) storage.Headers {
+	blocksByHeight := make(map[uint64]*flow.Block)
+	for _, b := range blocks {
+		blocksByHeight[b.Header.Height] = b
+	}
+
+	return synctest.MockBlockHeaderStorage(synctest.WithByHeight(blocksByHeight))
+}
+
+func (s *ScriptExecutorSuite) bootstrap() {
+	bootstrapOpts := []fvm.BootstrapProcedureOption{
+		fvm.WithInitialTokenSupply(unittest.GenesisTokenSupply),
+	}
+
+	executionSnapshot, out, err := s.vm.Run(
+		s.vmCtx,
+		fvm.Bootstrap(unittest.ServiceAccountPublicKey, bootstrapOpts...),
+		s.snapshot)
+
+	s.Require().NoError(err)
+	s.Require().NoError(out.Err)
+
+	s.height++
+	err = s.registerIndex.Store(executionSnapshot.UpdatedRegisters(), s.height)
+	s.Require().NoError(err)
+
+	s.snapshot = s.snapshot.Append(executionSnapshot)
+}
+
+func (s *ScriptExecutorSuite) SetupTest() {
+	s.log = unittest.Logger()
+	s.chain = flow.Emulator.Chain()
+
+	s.reporter = syncmock.NewIndexReporter(s.T())
+	s.indexReporter = index.NewReporter()
+	err := s.indexReporter.Initialize(s.reporter)
+	require.NoError(s.T(), err)
+
+	blockchain := unittest.BlockchainFixture(10)
+	s.headers = newBlockHeadersStorage(blockchain)
+	s.height = blockchain[0].Header.Height
+
+	entropyProvider := testutil.EntropyProviderFixture(nil)
+	entropyBlock := mock.NewEntropyProviderPerBlock(s.T())
+	entropyBlock.
+		On("AtBlockID", testifyMock.AnythingOfType("flow.Identifier")).
+		Return(entropyProvider).
+		Maybe()
+
+	s.snapshot = snapshot.NewSnapshotTree(nil)
+	s.vm = fvm.NewVirtualMachine()
```

Comment:
since you've written the code already, it's fine to keep it. In the future, you can simplify your tests using mocked interfaces. That may mean creating new interfaces or updating method arguments to accept an interface.

In this case, you could mock the `scriptExecutor *execution.Scripts`. There's already an [interface](https://github.com/onflow/flow-go/blob/192f26e669b02eea569914c0af607fb409553f6c/module/execution/scripts.go#L26).

you could also mock the version control object, so you can control the response precisely without a bunch of setup.

in general, I find this leads to easier to maintain and more robust unit tests since you have more contol.